File: javascript/src/routing.ts
import { ChatOpenAI } from "@langchain/openai";
import { StringOutputParser } from "@langchain/core/output_parsers";
import { traceable } from "langsmith/traceable";
import * as dotenv from "dotenv";
import path from "path";
import {
  BaseMessage,
  HumanMessage,
  SystemMessage,
} from "@langchain/core/messages";
import { PromptTemplate } from "@langchain/core/prompts";
import { z } from "zod";
dotenv.config({ path: path.resolve(__dirname, "../../.env") });

const model = new ChatOpenAI({
  modelName: "gpt-4o-mini",
  temperature: 0,
});

const parser = new StringOutputParser();

type AssistantInput = {
  id: string;
  description: string;
  prompt: string;
};

class Assistant {
  id: string;
  description: string;
  prompt: string;

  constructor(id: string, description: string, prompt: string) {
    this.id = id;
    this.description = description;
    this.prompt = prompt;
  }

  async run(input: string) {
    const messages: BaseMessage[] = [
      new SystemMessage({
        content: this.prompt,
      }),
      new HumanMessage(input),
    ];
    const chain = model.pipe(parser);
    const response = await chain.invoke(messages);
    return response;
  }

  static from(input: AssistantInput) {
    return new Assistant(input.id, input.description, input.prompt);
  }
}

type WorkflowInput = {
  assistants: Assistant[];
};

class Workflow {
  assistants: Assistant[];

  constructor(assistants: Assistant[]) {
    this.assistants = assistants;
  }

  private async _run(task: string) {
    const routerSchema = z.object({
      reason: z.string().describe("The reason for the choice"),
      assistant_id: z.string().describe(`The ID of the assistant to use. Choose from: ${this.assistants.map((assistant) => assistant.id).join(", ")}`),
    });
    const routerPrompt = PromptTemplate.fromTemplate(
      "Given a user prompt/query: {user_query}, select the best option out of the following routes:\n{routes}\nAnswer only in JSON format."
    );
    const routes = this.assistants
      .map((assistant) => `ID: ${assistant.id}, Description: ${assistant.description}`)
      .join("\n");
    const routerChain = routerPrompt.pipe(
      model.withStructuredOutput(routerSchema, {
        strict: true,
      })
    );
    const response = await routerChain.invoke({
      user_query: task,
      routes,
    });
    const assistant = this.assistants.find(
      (assistant) => assistant.id === response.assistant_id
    );
    if (!assistant) {
      throw new Error(`Assistant with ID ${response.assistant_id} not found`);
    }
    const result = await assistant.run(task);
    return result;
  }

  async run(task: string) {
    return traceable(this._run.bind(this), { name: "routing" })(task);
  }

  static from(input: WorkflowInput) {
    return new Workflow(input.assistants);
  }
}

async function main() {
  const code_generation = Assistant.from({
    id: "code_generation",
    description: "Suited for code generation",
    prompt: "You are a helpful assistant that generates code for a website",
  });
  const trip_planner = Assistant.from({
    id: "trip_planner",
    description: "Suited for trip planning",
    prompt: "You are a helpful assistant that plans trips",
  });
  const story_teller = Assistant.from({
    id: "story_teller",
    description: "Suited for story telling",
    prompt: "You are a helpful assistant that tells stories",
  });
  const workflow = Workflow.from({ assistants: [code_generation, trip_planner, story_teller] });
  const tasks = [
    "Write a Python function to check if a number is prime.",
    "Plan a 2-week trip to Europe.",
    "Tell me a story about a cat.",
  ];
  for (const task of tasks) {
    const result = await workflow.run(task);
    console.log(result);
  }
}

main();

File: python/langchain/routing.py
# https://www.agentrecipes.com/routing
# Conditional Router Agent Workflow
from typing import Annotated
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_openai import ChatOpenAI
from typing import TypedDict, List
import asyncio
from langsmith import traceable
from dotenv import load_dotenv

load_dotenv()

model = ChatOpenAI(model="gpt-4o-mini", temperature=0)


class Assistant:
    """An assistant is a model that can be used to solve a task."""

    id: str
    description: str
    system_prompt: str

    def __init__(self, id: str, description: str, system_prompt: str):
        self.id = id
        self.description = description
        self.system_prompt = system_prompt

    async def run(self, input_query: str) -> str:
        messages = [("system", self.system_prompt), ("user", input_query)]
        parser = StrOutputParser()
        chain = model | parser
        response = await chain.ainvoke(messages)
        return response


class RouterSchema(TypedDict):
    """A schema for the router workflow."""

    reason: Annotated[str, ..., "The reason for selecting the route"]
    route: Annotated[
        str, ..., "The route selected. Must be one of the ids in the list of routes."
    ]


class RouterWorkflow:
    """A workflow that routes a task to the assistant best suited for the task."""

    def __init__(self, assistants: List[Assistant]):
        self.assistants = assistants

    @traceable(name="routing")
    async def run(self, input_query: str) -> str:
        """Given a `input_query` and a dictionary of `routes` containing options and details for each.
        Selects the best route for the task and return the response from the model.
        """
        ROUTER_PROMPT = """Given a user prompt/query: {user_query}, select the best option out of the following routes:
        {routes}. Answer only in JSON format."""
        prompt = PromptTemplate.from_template(ROUTER_PROMPT)
        model_routes_str = "\n".join(
            [f"id: {v.id}, description: {v.description}" for v in self.assistants]
        )
        chain = prompt | model.with_structured_output(
            RouterSchema, method="json_schema", strict=True
        )
        response = await chain.ainvoke(
            {"user_query": input_query, "routes": model_routes_str}
        )
        profiles_dict = {profile.id: profile for profile in self.assistants}
        assistant = profiles_dict[response["route"]]
        if not assistant:
            raise ValueError(f"Assistant with id {response['route']} not found")
        return await assistant.run(input_query)


async def main():
    assistants = [
        Assistant(
            id="code_generation",
            description="Suited for code generation",
            system_prompt="You are a helpful assistant that generates code for a website",
        ),
        Assistant(
            id="trip_planner",
            description="Suited for trip planning",
            system_prompt="You are a helpful assistant that plans trips",
        ),
        Assistant(
            id="story_teller",
            description="Suited for story telling",
            system_prompt="You are a helpful assistant that tells stories",
        ),
    ]
    tasks = [
        "Write a Python function to check if a number is prime.",
        "Plan a 2-week trip to Europe.",
        "Write a story about a brave knight and a dragon.",
    ]
    router = RouterWorkflow(assistants)
    responses = await asyncio.gather(*[router.run(task) for task in tasks])
    return responses


if __name__ == "__main__":
    response = asyncio.run(main())
    print(response)
